### ISSUE TITLE
Papa games issues

### ISSUE THREAD

**Issue by MartySVK** [ruffle-rs/ruffle#9201]
### Describe the bug

Since Papa games are very popular, I decided to make a special issue for them and test them regularly.

All games use similar logic and will also contain similar features. In addition to Papa games, Flipline Studios has similar games like Jacksmith.

Current issues:
- [ ] loading bar bug
- [ ] timer-volume  error
- [ ] addEventListener error
- [ ] maxChars error

From the discussion on discord, maxChars could be stubbed. Loading bar is probably just a visual issue and not every Papa game has this problem. The game is still loading, but you can click on "Loading" text and continue to the menu.

Error Logs:

`Error dispatching event EventObject(EventObject { type: "timer", class: flash.events::TimerEvent, ptr: 0x116ae670 }) to handler FunctionObject(FunctionObject { ptr: 0x11a02f10 }) : RustError("Cannot access property ::volume of null or undefined")`

`Error dispatching event EventObject(EventObject { type: "addedToStage", class: flash.events::Event, ptr: 0x1206cb48 }) to handler FunctionObject(FunctionObject { ptr: 0x12065480 }) : RustError("Cannot access property ::addEventListener of null or undefined")`

`Error dispatching event EventObject(EventObject { type: "enterFrame", class: flash.events::Event, ptr: 0x11f84410 }) to handler FunctionObject(FunctionObject { ptr: 0x11a1d620 }) : RustError("Cannot set undefined property ::maxChars on Some(flash.text::TextField)")`

### Expected behavior

Click the PLAY button to continue the game

### Affected platform

Desktop app

### Operating system

Windows 10

### Browser

Chrome 109

**Comment by MartySVK** [comment]
> You should also mention the unclickable menu buttons in games like Papa's Burgeria on Flipline.com: https://www.flipline.com/games/papasburgeria/index.html
> 
> `ERROR core/src/avm2/events.rs:422 Error dispatching event EventObject(EventObject { type: "addedToStage", class: flash.events::Event, ptr: 0x1a24058 }) to handler FunctionObject(FunctionObject { ptr: 0x6d8bac8 }) : RustError("Cannot access property ::addEventListener of null or undefined")`

Actually, these kinds of issues are mostly connected with different issues. In this case, it is connected with Stage invalid. I tried @golfinq PR (https://github.com/ruffle-rs/ruffle/pull/8561) and you can click on the menu button, you can even reach the first customer 🥳. Hope, it will be merged soon

**Comment by n0samu** [comment]
With #9339 merged, Papa's Burgeria works until here, where for some reason it's impossible to click the "Take Order" button:
![image](https://user-images.githubusercontent.com/71368227/217714677-6845e84d-9ffc-43b2-881e-964f7cbe4c6a.png)

I'm guessing this is a mouse pick issue, but it needs further investigation.

**Comment by MartySVK** [comment]
yeah, same for me. I will check other papa games later and add error details

**Comment by Dinnerbone** [comment]
Papa's Hot Doggeria seems to need `BitmapData.hitTest` to progress that screen

**Comment by MartySVK** [comment]
wow...nice progress

**Comment by Lord-McSweeney** [comment]
As part of the graphical issues, Jacksmith seems to have red circles covering weapons (I believe where they were hit with the hammer). Jacksmith finally breaks on `ERROR core/src/display_object/movie_clip.rs:1556 Unable to instantiate display node id 1247, reason being: Not a DisplayObject`, after delivering the sword to the first customer.

**Comment by n0samu** [comment]
With #9683 rebased on latest master, Papa's Cupcakeria works until you try to frost the cupcake:
![image](https://user-images.githubusercontent.com/71368227/221242639-bdeb9d1b-6092-41f1-a736-fae6e9303791.png)
The error is:
```
ERROR run_frame:run_all_phases_avm2: ruffle_core::avm2::activation: ReferenceError: Error #1069: Property getVector not found on flash.display.BitmapData and there is no default value.
        at cakeGame.models::CupcakeFrosting/getFrostingY()
        at cakeGame.models::CupcakeFrosting/renderFrostingQueue()
        at cakeGame.models::CupcakeFrosting/updateObject()
        at cakeGame.models::Cupcake/updateBuildCupcake()
        at cakeGame.models::FoodOrder/updateBuildCupcakes()
        at cakeGame.screens::BuildScreen/updateScreen()
        at cakeGame.events::GameControls/gameCycle()
```

**Comment by Dinnerbone** [comment]
I had no idea there were so many yet unsupported BitmapData methods 😅

**Comment by n0samu** [comment]
Papa's Pastaria works up until you try to pour sauce onto the pasta. Seems like a similar issue to Hot Doggeria.
![image](https://user-images.githubusercontent.com/71368227/221245577-fd601277-39cd-49d6-bdeb-7c3847f7c899.png)

**Comment by n0samu** [comment]
Papa's Donuteria is officially the first Papa's game to make it through the tutorial! 🎉 
![image](https://user-images.githubusercontent.com/71368227/221247308-a58397a9-e2ce-4f18-a6c7-8b4eb7bc0133.png)

Unfortunately, the "Take Order" button is unclickable here, even though it was clickable when starting the tutorial.

**Comment by n0samu** [comment]
Papa's Bakeria is in exactly the same situation as Donuteria:
![image](https://user-images.githubusercontent.com/71368227/221254525-d978a670-51b9-447d-aa72-9c7319ba0574.png)

Papa's Cheeseria gets stuck when trying to add turkey to the sandwich, with the same BitmapData.getVector error as Cupcakeria:
![image](https://user-images.githubusercontent.com/71368227/221254400-58addc88-47de-4207-a2ef-09172414e820.png)

Papa's Sushiria gets stuck here:
![image](https://user-images.githubusercontent.com/71368227/221251497-fc8f298a-8afd-4725-b0d4-62e2cfd55668.png)
The error is:
```
ERROR ruffle_core::avm2::activation: ReferenceError: Error #1069: Property threshold not found on flash.display.BitmapData and there is no default value.
        at sushiGame.models::SushiTopping/cutoutDrop()
        at sushiGame.models::SushiTopping/addDrop()
        at sushiGame.screens::BuildScreen/releaseToppingBin()
        at sushiGame.screens::BuildScreen/releaseScreen()
```

Papa's Scooperia gets stuck after trying to pour the strawberry syrup:
![image](https://user-images.githubusercontent.com/71368227/221253805-ca9c9c87-ba03-47dc-9bd2-a2a17ceec2c1.png)

The error is:
```
ERROR ruffle_core::avm2::activation: ReferenceError: Error #1069: Property drawPath not found on flash.display.Graphics and there is no default value.
        at scoopGame.models::SundaeTopping/screenUpdate()
        at scoopGame.models::FoodOrder/updateOnBuildScreen()
        at scoopGame.screens::BuildScreen/updateScreen()
        at scoopGame.events::GameControls/gameCycle()
```

**Comment by MartySVK** [comment]
@n0samu Thanks for the deep testing. I wanted to do the same thing tomorrow, so after the next update I will definitely do similar tests of all games.

**Comment by MartySVK** [comment]
With PR https://github.com/ruffle-rs/ruffle/pull/9683, in Jacksmith I can finish my first order (tutorial), but it is not possible to take another

![image](https://user-images.githubusercontent.com/10010569/222007116-a8fd8680-93bd-4652-a93f-28553117eebf.png)


``2023-02-28T23:27:56.131047Z ERROR ruffle_core::avm2::activation: AVM2 error: Object(ErrorObject(ErrorObject { class: TypeError, ptr: 0x1d62b68a8e0 }))
2023-02-28T23:27:56.131459Z ERROR ruffle_core::avm2::activation: TypeError: Error #1009: Cannot access a property or method of a null object reference. (accessing field: ::addEventListener)
2023-02-28T23:27:56.131825Z ERROR ruffle_core::avm2::events: Error dispatching event EventObject(EventObject { type: "addedToStage", class: flash.events::Event, ptr: 0x1d62b421da0 }) to handler FunctionObject(FunctionObject { ptr: 0x1d62b175e00 }) : AvmError(Object(ErrorObject(ErrorObject { class: TypeError, ptr: 0x1d62b68a8e0 })))
2023-02-28T23:28:01.751692Z ERROR ruffle_core::display_object::movie_clip: Unable to instantiate display node id 1247, reason being: Not a DisplayObject``

**Comment by MartySVK** [comment]
After Papa's Cupcakeria is blocked by unimplemented Graphics.lineBitmapStyle
```2023-03-02T23:35:19.142079Z ERROR ruffle_core::avm2::activation: AVM2 error: Object(ErrorObject(ErrorObject { class: ReferenceError, ptr: 0x182e3c827f0 }))
2023-03-02T23:35:19.142903Z ERROR ruffle_core::avm2::activation: ReferenceError: Error #1069: Property lineBitmapStyle not found on flash.display.Graphics and there is no default value.

**Comment by MartySVK** [comment]
@n0samu, can you please test Bakeria again? I can not finish the tutorial because of `panicked at 'assertion failed: self.mask_index_count.is_none()', render\src\tessellator.rs:127:17`

**Comment by n0samu** [comment]
That's strange. I just ran through the tutorial with the latest desktop nightly and did not see that issue.

**Comment by MartySVK** [comment]
you are right...it is probably only in debug

**Comment by c4artisan** [comment]
by the pull request that's above this comment taco mia (and i think burgeria but i haven't tested it yet) works almost flawlessly
however the taco toppings go to the bottom of the shell and (of course) there are no effects
![image](https://user-images.githubusercontent.com/41709262/223889132-349c9444-bf7e-4c80-9cf3-487bd413779b.png)

**Comment by c4artisan** [comment]
the errors on the console:
![image](https://user-images.githubusercontent.com/41709262/223889644-7a962fd9-3508-4864-9544-4057c9fdc102.png)

**Comment by MartySVK** [comment]
> Hot Doggeria

it is probably because of hitTest stub. Same for Hot Doggeria

**Comment by c4artisan** [comment]
ah. hope it's implemented soon

**Comment by InputEvelution** [comment]
The issue of toppings not piling up correctly also seems to be an issue for Papa's Freezeria. Freezeria also seems to have a strange sound problem when switching between certain scenes, where the audio will suddenly loudly peak for a short moment.

**Comment by InputEvelution** [comment]
Pancakeria seems to have similar sound problems to Freezeria. The text and font rendering issues pose a large problem for this game, as it's impossible to see how many of a certain topping you need to put on from the ticket. Additionally, the ticket read-off is bugged; past the first detail being listed by the customer, everything else will appear on the ticket in a split second.

**Comment by InputEvelution** [comment]
While deleting save game data works fine for Burgeria currently, functionality seems to be broken in Taco Mia onwards.

**Comment by n0samu** [comment]
Wingeria also has some noticeable font issues.

**Comment by MartySVK** [comment]
In Hot Drogeria, when putting toppings on the hotdog, only the first topping that touches it seems to stick, the rest just fall down.
![image](https://user-images.githubusercontent.com/10010569/225586869-348c62f6-943e-4e63-abc0-a7d1b2ba0627.png)

**Comment by c4artisan** [comment]
> In Hot Drogeria, when putting toppings on the hotdog, only the first topping that touches it seems to stick, the rest just fall down. ![image](https://user-images.githubusercontent.com/10010569/225586869-348c62f6-943e-4e63-abc0-a7d1b2ba0627.png) left - Ruffle right - Flash Player
> 
> But mustard sticks without any problems

:< i hope that doesn't happen in taco mia

**Comment by c4artisan** [comment]
it does. let's just hope it's fixed soon

**Comment by n0samu** [comment]
Unfortunately the games have regressed on desktop due to an XML issue - see #10498

**Comment by n0samu** [comment]
The XML issue will be fixed by #10500.

Playing through the tutorial in Freezeria, I noticed a similar error to the one in Pastaria, although this one just causes a tiny glitch while the first customer is ordering, with seemingly no effect on gameplay.
`ERROR run_frame:run_all_phases_avm2: ruffle_core::avm2::events: Error dispatching event EventObject(EventObject { type: "enterFrame", class: flash.events::Event, ptr: 0x1a8f83ac070 }) to handler FunctionObject(FunctionObject { ptr: 0x1a8e2532d10 }) : RustError("ArgumentError: sprinkles is not a valid frame label.")`

The issues with missing text in Freezeria, Pancakeria, and Wingeria all seem to be fixed by #10494! 🎉

**Comment by InputEvelution** [comment]
> The issues with missing text in Freezeria, Pancakeria, and Wingeria all seem to be fixed by #10494! 🎉

Are you sure? Pancakeria still lacks the text indicating how many of a topping to put on with the tickets for me.

**Comment by n0samu** [comment]
Oh, you're right. I'm not sure whether that's really a font issue though. While the customer is ordering, there are several errors, and the ticket is drawn faster than it's supposed to be:
`ERROR run_frame:run_all_phases_avm2: ruffle_core::avm2::events: Error dispatching event EventObject(EventObject { type: "enterFrame", class: flash.events::Event, ptr: 0x20a880e6390 }) to handler FunctionObject(FunctionObject { ptr: 0x20a84b6ff10 }) : RustError("ArgumentError: pancake is not a valid frame label.")`
`ERROR run_frame:run_all_phases_avm2: ruffle_core::avm2::events: Error dispatching event EventObject(EventObject { type: "enterFrame", class: flash.events::Event, ptr: 0x20a8810d990 }) to handler FunctionObject(FunctionObject { ptr: 0x20a84b6ff10 }) : RustError("ArgumentError: butter is not a valid frame label.")`
`ERROR run_frame:run_all_phases_avm2: ruffle_core::avm2::events: Error dispatching event EventObject(EventObject { type: "enterFrame", class: flash.events::Event, ptr: 0x20a8810e110 }) to handler FunctionObject(FunctionObject { ptr: 0x20a84b6ff10 }) : RustError("ArgumentError: blueberries is not a valid frame label.")`

**Comment by MartySVK** [comment]
unfortunately Pastaria is broken with similar error

**Comment by c4artisan** [comment]
any updates on the hittest thing?

**Comment by InputEvelution** [comment]
Some interesting observations about the current issues with ingredients going to the bottom in Taco Mia:

- Toppings have no problems colliding with themselves - if you pour beans, the beans will pile up just fine against one another.
- The issue of only the first bit of topping sticking to the taco also seems to apply with toppings - if you pour a pile of beans, and then a pile of onions, only the first bit of onions will stick to the pile of beans. This doesn't happen if you pour the same type of topping twice in a row, they'll stack up against one another just fine.
- Sauces will stack up against whatever the last type of topping you added was - if you last poured a bunch of tomato, it'll go on the tomato. If you didn't pour any non-sauce toppings yet, it'll land on the taco just fine.

All in all, it seems the issue may be that ruffle's hitTest is only recognising the most recent type of object to be added to the taco's collision mesh. The moment that changes, everything that came before is immediately ignored.

**Comment by InputEvelution** [comment]
> While the customer is ordering, there are several errors, and the ticket is drawn faster than it's supposed to be.

This seems to be a problem in Freezeria as well, though not to the same extent - it's only when the ticket ordering gets to the last few items (such as sprinkles) that it seems to start going faster than intended.

**Comment by InputEvelution** [comment]
An obscure issue with Papa's Pizzeria (that primarily affects TASing) has emerged: https://github.com/ruffle-rs/ruffle/issues/10682

**Comment by InputEvelution** [comment]
The Breakfast Blast minigame in Pancakeria and Wingeria seems to be completely broken currently - selecting Play will simply open the curtains up to a black screen, with the player unable to do anything other than close and re-open Ruffle.

**Comment by InputEvelution** [comment]
Donuteria has an odd visual bug where when inserting fillings into the donuts, the tube connecting the end to the wall will be completely invisible. Gameplay-wise, however, all appears to work as intended so far.

**Comment by InputEvelution** [comment]
All the gamerias from Wingeria onwards allow the player to take a photo of their character in-game and save it to their computer. When clicking the "Save Photo" button, Ruffle gets stuck and won't respond to input without some finicky workarounds (clicking off of Ruffle and back in seems to sometimes fix the problem).

**Comment by n0samu** [comment]
Save Photo (and saving/loading backup save files) needs FileReference methods that Ruffle hasn't implemented yet.

**Comment by n0samu** [comment]
As mentioned in #10760, some of the button hover effects are incorrect. You can see it immediately in Burgeria as well:
![burgeriabuttons](https://user-images.githubusercontent.com/71368227/233591331-48bf7d62-5721-4d29-b80e-b7b8631b4caf.png)

These buttons have "over" state records with the Overlay blend mode. For some reason this isn't working in Ruffle. Maybe button records with blends don't have special handling implemented yet.

Here is an isolated SWF of just the HUD buttons from Pancakeria, for easier testing:
[PancakeriaHUDButtons.zip](https://github.com/ruffle-rs/ruffle/files/11294090/PancakeriaHUDButtons.zip)

**Comment by InputEvelution** [comment]
Here's a list of graphical issues that have come to my attention in testing Cheeseria and Bakeria recently:

- Both games seem to have issues handling colours on the chef and server's shoes at various points, causing them to appear white and grey instead. Turning on RUST_LOG=warn causes the console to complain about not being able to find childs with names such as "color1" and "color2" at such points, which I presume to be related. 
![ShoesNoColour](https://user-images.githubusercontent.com/29367840/233837085-88f4befc-ab49-4748-a67c-7728c72ca06f.PNG)

- Pouring Streusel topping onto a pie in Bakeria causes its graphics to become corrupted along the right columns of pixels. 
![PieCorruption2](https://user-images.githubusercontent.com/29367840/233837098-6d03d65c-d88e-4f0d-ac13-b1310e5022e6.PNG)

- The text in the graphics for Special Recipe lacks any highlight effects, making them incredibly hard to read in various situations.
![SpecialsText](https://user-images.githubusercontent.com/29367840/233837105-fd4695c3-b8e2-46d5-af3e-0666fb8d99d7.PNG)

- The background shopping centre typically visible from the counter in Bakeria is completely invisible, with a black background taking its place.
![BrokenBackground](https://user-images.githubusercontent.com/29367840/233837112-6092a9c0-cb2b-4608-aa7c-8019571f7d4d.PNG)

- The Papa's Raceway minigame found in Donuteria, Cheeseria, Bakeria and Sushiria suffers from layering issues - karts will either always be rendered in front or always be rendered behind other objects instead of shifting dynamically like in Flash.
![RacewayLayering](https://user-images.githubusercontent.com/29367840/233837118-39a37d8e-2080-47cd-8c59-c2bd216f0411.PNG)

- The waterfall graphics become corrupted at the end of a day in Bakeria, with columns of black pixels moving to the left every second or so. (And as with other games, the floor right outside the Bakeria isn't rendered correctly.)

**Comment by InputEvelution** [comment]
Sushiria seems to have two gameplay issues in Ruffle that I haven't seen mentioned here:

- In the Tea Station, plates of sushi seem to have abnormally high reaching hitboxes - attempting to click on the tea will instead lead to the player clicking on the plate of sushi unless they click the lid of the tea or slightly below it.

- In the Build Station, pouring any sauce off the side of the sushi whatsoever will result in drastically lower scores for the station, with penalties ranging from 30% to as much as 90% knocked off an otherwise flawless build station performance. This is particularly problematic, as it leaves extremely little room for error on the player's part (especially with closers), and exacerbates Sushiria's already notorious difficulty.

**Comment by lemaximedu66** [comment]
> Save Photo (and saving/loading backup save files) needs FileReference methods that Ruffle hasn't implemented yet.

still not implemented yet as of today

**Comment by c4artisan** [comment]
anything on the taco mia and hot doggeria issues? (the topping ones)

**Comment by n0samu** [comment]
The invisible frosting in Scooperia has been fixed by #11898. And button blends & filters are fixed by #11891. (Besides a minor issue that I described in #11907)

**Comment by MartySVK** [comment]
I noticed new regression in Pastaria, when I use tomato sauce, the texture is invisible. It will become visible at the end. This was working before.
![image](https://github.com/ruffle-rs/ruffle/assets/10010569/5a12f53a-6939-422f-a573-875466a87956)
![image](https://github.com/ruffle-rs/ruffle/assets/10010569/61c77f99-c712-47ab-91d1-033334c8ea31)

**Comment by n0samu** [comment]
I found another rendering regression caused by cacheAsBitmap: #12030

**Comment by InputEvelution** [comment]
Checking the latest nightly, it seems that the issues with orders being written out too quickly has been fixed. The Breakfast Blast minigame also no longer hangs, though the graphics and gameplay still seem to be broken. I noticed a new issue with Pancakeria this time though where certain graphics would appear glitched or corrupted, most notably while pouring sauces onto pancakes and during the tutorial.

The issues with topping collisions not working properly in Taco Mia, Freezeria and Hot Doggeria still hold true, as do the brief loud sounds when moving between scenes in Freezeria onwards.

**Comment by InputEvelution** [comment]
Other issues that seem to have been fixed in latest nightly:
- Tomato sauce regression in Pastaria
- Filling tubes in Donuteria not appearing
- Layering issues in Papa's Raceway minigame
- Sauces etc. missing graphical effects
- Scoring in Sushiria

Other new issues:
- Pouring sauces onto pies in Bakeria can sometimes cause graphics to become corrupted
- Sauces etc. seem to wiggle a bit vertically while being poured onto food (Pancakeria and Bakeria come to mind as examples)

**Comment by n0samu** [comment]
I'm not seeing the corrupted graphics in Pancakeria, can you show me what you're talking about? And does it happen in both desktop and web builds of Ruffle?

**Comment by InputEvelution** [comment]
> I'm not seeing the corrupted graphics in Pancakeria, can you show me what you're talking about? And does it happen in both desktop and web builds of Ruffle?

Checking on experimonkey, it does indeed seem to be limited to the desktop builds of Ruffle. Here's a couple videos showing what I'm talking about:

https://github.com/ruffle-rs/ruffle/assets/29367840/6e0cb53a-769a-4515-9636-3505ce0b8cc7

https://github.com/ruffle-rs/ruffle/assets/29367840/5da45832-8a61-477d-895f-b4ac298a8b4a

**Comment by n0samu** [comment]
Your first video honestly looks to me like some kind of driver bug. Your second video doesn't seem to show corrupted graphics, but just "jittery" graphics similar to what I pointed out in my 3 most recent issue reports. Issues with jittery graphics seem to be related to pixel snapping.

**Comment by InputEvelution** [comment]
You can see corrupted graphics in the second video, briefly appearing a couple of times along the right side above where the blueberry sauce is being poured. This is a different issue to the jitter.

**Comment by InputEvelution** [comment]
> Your first video honestly looks to me like some kind of driver bug.

This has now been confirmed to be the issue - I updated my NVIDIA drivers and the issues were solved 👍

**Comment by InputEvelution** [comment]
Papa's Donuteria seems to have a really obscure issue: If the player tries to click on one of the drizzles **within the first 5 frames of them appearing onscreen during the tutorial**, Ruffle will crash with this error message:
![DonuteriaDrizzleEarlyError](https://github.com/ruffle-rs/ruffle/assets/29367840/5eb6685c-0beb-42b6-94a4-022dd0d334ab)
The console provides a little bit more information:
![DonuteriaDrizzleEarlyError2](https://github.com/ruffle-rs/ruffle/assets/29367840/00cca276-3e09-4ae7-a80d-b2facce3fefd)
This does not happen outside of the tutorial, and clicking the drizzles any later works normally.

Additionally, while I had initially assumed that the heavy stuttering present in the Papa Louie Arcade intro from Pancakeria onwards was a performance issue, running the game in libTAS suggests that there may be something else going on - it seems to follow an extremely consistent rhythm, rather than being dependent on the user's hardware. Trying to set a custom framerate in ruffle also seems to affect the manner in which it plays out.

**Comment by c4artisan** [comment]
![image](https://github.com/ruffle-rs/ruffle/assets/41709262/348d893b-1e61-409f-a051-4fcaa4c4b0de)
the taco mia ingredient issue is finally fixed!

**Comment by c4artisan** [comment]
and so is the hot doggeria issue.

**Comment by MartySVK** [comment]
All major bugs have been fixed. If the game contains any glitches or other issues, it will be better to make a separate issue.
